docs/cmdline-opts/type-of-service.md

```diff
@@ -0,0 +1,56 @@
+---
+c: Copyright (C) Daniel Stenberg, <daniel@haxx.se>, et al.
+SPDX-License-Identifier: curl
+Long: type-of-service
+Arg: <string>
+Help: Configure IP Type of Service (TOS)
+Added: 8.9.0
+Category: connection
+Protocols: All
+Multi: single
+See-also:
+  - tcp-nodelay
+Example:
+  - --type-of-service CS5 $URL
+---
+
+# `--type-of-service`
+
+Set IP Type of Service (TOS). (Added in 8.9.0).
```

Comment:
Since you added  support for IPv6 and its "Traffic Class" field:

1. Should this option really be named like this now?
2. Should the documentation not specify the IPv6 details?

Reply:
1. Do you have a better suggestion? Maybe add another `--traffic-class` which does the same?
2. These fields have the same purpose and structure (6 bits DSCP, 2 bits ECN), only the name and the location within the packet differs.

Reply:
No, this should be a single option I think. Having two different ones doing the same thing does not make anything easier for anyone. It would make more sense to come up with a generic name that works for applying to both cases. Maybe it should even include "IP" in the name to better hint what it is about.

Maybe...

- `--ip-traffic`
- `--ip-type`
- `--ip-service`

?

Reply:
I'm not sure any of these suggestions is more descriptive than type-of-service.

TOS used to have a different meaning, but nowadays, it is just like traffic class, divided to 6 bits DSCP and 2 bits ECN.

Maybe we should just have `--ip-dscp` and `--ip-ecn`?

Reply:
Just some helpful info on other tools:

linux iproute2 has `tos` and its alias `dsfield` for adding/showing routes that match on the dsfield. It accepts a named DSCP value or a masked 8 bit TOS value, and the ECN bits are ignored. The option is named the same for both v4 and v6 routing tables. E.g. `ip route add to 1.2.3.4/32 dsfield AF21 counter`

linux nftables has `ip dscp` and `ip6 dscp` payload expresions that match/set the 6 bit ds field in the respective protocol family. It accepts either a named or 6 bit DSCP value. E.g. `nft add rule inet filter input ip dscp af21 meta priority set 1:4`.

`tcpdump -v` displays the tos / tclass byte as "tos" or "class" followed by the 8 bit hex value of the byte, including the ECN field.

`ping -Q` sets the tos / tclass byte just like this patch, and linux copies the entire byte in echo replies.

`--ip-tos` or even just `--tos` seems fine to me. I'm partial to `--dscp` I think. I mentioned it earlier but  linux ignores the ECN field from this sockopt for SOCK_STREAM sockets, and I suppose the quic library will manage the ECN field otherwise, so I expect that setting the the ECN field with this option will never be effective, hence `--dscp` seems reasonable.

Reply:
@rpigott Thanks a lot for the detailed overview!

I'm a bit reluctant. My considerations:
* `tos`/`type-of-service`:
  * Pro: Widely used (relatively for this niche feature :)) and should be easily recognized.
  * Con: It's an obsolete name for this field, and doesn't describe its actual use.
* `traffic-class`:
  * Pro: Better description of the field, was chosen over ToS for IPv6, which is more recent.
  * Con: Less known, harder to discover.

I call for a vote.

❤ `--ip-tos`
🚀 `--tos`
🎉 `--ip-type-of-service`
👀 `--ip-traffic-class`

Reply:
Aaaaaaand we have a winner!

Changed to `--ip-tos`. Ready for review.